llext: convert "modules" tests to sample

pillo79 opened the pull request:
The current llext test suite contains a number of tests that are called `modules_enabled` and were originally added in #67105 as a way to showcase the Kconfig `tristate` element type; those have since been expanded to 8 tests covering all current llext option combinations.

This abundance of tests, however, is not useful: since the different Kconfig options have no practical effect on the llext code, all those tests perform exactly like the non-modules-enabled configuration and just double the testing effort and number of configurations in `testcase.yaml`. 

This PR replaces those redundant tests with a sample that shows how to actually create a "module" in the Linux kernel sense - a piece of code that can be either linked in or loaded at runtime.

pillo79 committed:
llext: add a "modules" Kconfig tristate example

This adds a new sample to demonstrate the use of tristate symbols
in Kconfig to build a function as an llext module or as a built-in
part of Zephyr.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
llext: remove the "modules_enabled" tests

These tests are simple copies of the existing tests with the addition of
the CONFIG_MODULES=y option. This different Kconfig setting has no
practical effect on the code, and the tests are therefore redundant.

Remove them to halve the number of tests.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

teburd reviewed (dismissed):
Thank you for doing this, and yes this makes great sense to not repeat the same work over and over for showing how the tristate Kconfig works.

@lyakh should run this with qemu_xtensa at the very least and approve before merging, marking DNM (with my +1 here) until @lyakh does just that. Feel free to remove DNM once reviewed and approved by @lyakh.

kartben commented on samples/subsys/llext/modules/README.rst:
```diff
@@ -0,0 +1,59 @@
+their symbols can be accessed and functions called.
+
+Specifically, this shows how to call a simple "hello world" function,
+implemented in ``src/hello_world_ext.c``. This is achieved in two different ways,
```
nit (i.e. not worth updating unless there's another round of reviews needed): maybe make this a `:zephyr_file:` like you're doing for others

mathieuchopstm commented on tests/subsys/llext/simple/testcase.yaml:
```diff
@@ -36,50 +36,6 @@ tests:
-      - simulation
-      - arch
-    platform_exclude:
-      - qemu_cortex_a9 # MMU
```
Cortex-A9 platform exclusion seems gone in `sample.yaml`.
Is this on purpose?

pillo79 replied:
Yes, that exclusion was there before the runtime filter was added, so that platform should be caught by the above filter anyway. I will double-check this before submitting v2 though.

kartben commented on samples/subsys/llext/modules/README.rst:
```diff
@@ -0,0 +1,59 @@
+   :relevant-api: llext
+
+    Call a function in a loadable extension module,
+    either built in or loaded at runtime.
```
```suggestion
    either built-in or loaded at runtime.
```

kartben commented on samples/subsys/llext/modules/Kconfig:
```diff
@@ -0,0 +1,18 @@
+	default m
+	help
+	  This enables calling the hello_world function, implemented in
+	  hello_world_ext.c, either as an llext module or as a built in part of
```
```suggestion
	  hello_world_ext.c, either as an llext module or as a built-in part of
```

mathieuchopstm commented on samples/subsys/llext/modules/sample.yaml:
```diff
@@ -0,0 +1,34 @@
+  filter: not CONFIG_MPU and not CONFIG_MMU and not CONFIG_SOC_SERIES_S32ZE
+  platform_exclude:
+    - apollo4p_evb
+    - apollo4p_blue_kxr_evb
```
Please add a comment referring to issues 72775 / 73443 for apollo platforms, to make the reason why they are excluded obvious (as done for the numaker right below).

lyakh commented on samples/subsys/llext/modules/sample.yaml:
```diff
@@ -0,0 +1,34 @@
+tests:
+  sample.llext.modules_enabled:
+    extra_configs:
+      - CONFIG_MODULES=y
```
just to make sure this is correct - it's also set in prj.conf and left unchanged below in the monolithic case - is that correct?

pillo79 replied:
Thanks, you are right, it's an useless line. The point is to have `MODULES` always enabled during this test so I will remove this.

kartben reviewed (dismissed):
this is awesome stuff - thanks! Some minor nits/typos in case this goes through another round of updates.

mathieuchopstm commented on samples/subsys/llext/modules/src/hello_world_ext.c:
```diff
@@ -0,0 +1,21 @@
+{
+	printk("Hello, world!\n");
+}
+LL_EXTENSION_SYMBOL(hello_world);
```
If `CONFIG_LLEXT_HELLO_WORLD=y`, this file will be built as part of the Zephyr application. However, there is nothing special in the `LL_EXTENSION_SYMBOL` macro that makes it not emit the export table entry.
Is there no risk to blow up CI with orphan section linker warnings due to something being emitted to the `.exported_sym` section which, AFAICT, is not handled in Zephyr linker script?

pillo79 replied:
Interesting... it did not in my tests, but I can't be sure it won't everywhere. I will fix this as we did for the `EXPORT_SYMBOL` macro: make it a NOP if LLEXT is not enabled, and disable LLEXT for the =y case.

mathieuchopstm commented on samples/subsys/llext/modules/prj.conf:
```diff
@@ -0,0 +1,9 @@
+CONFIG_LOG=y
+CONFIG_LOG_MODE_IMMEDIATE=y
+
+CONFIG_MODULES=y
```
Nit, and definitely out of this PR's scope: I find this Kconfig option's name quite confusing at first glance.

pillo79 replied:
Indeed it is. We can't change it unless we modify `kconfiglib` internals, though, since that is the name that triggers the whole "tristate" stuff in Kconfig. :slightly_frowning_face:

mathieuchopstm commented on samples/subsys/llext/modules/src/main_builtin.c:
```diff
@@ -0,0 +1,23 @@
+LOG_MODULE_REGISTER(app);
+
+#include <zephyr/llext/llext.h>
+#include <zephyr/llext/buf_loader.h>
```
These includes seem unneeded.

lyakh commented on tests/subsys/llext/simple/testcase.yaml:
```diff
@@ -36,50 +36,6 @@ tests:
-    extra_configs:
-      - CONFIG_MODULES=y
-      - CONFIG_LLEXT_STORAGE_WRITABLE=y
-      - CONFIG_LLEXT_TYPE_ELF_RELOCATABLE=y
```
as an example, I looked for another `CONFIG_LLEXT_STORAGE_WRITABLE=y && CONFIG_LLEXT_TYPE_ELF_RELOCATABLE=y` test here since you're saying, that you're removing copies, and I haven't found such a test, am I missing something?

pillo79 replied:
No, you're right - that was a proper mistake. :slightly_smiling_face: 
I will make sure to check all removed paths and convert them to non-MODULES in case they were not duplicates.

mathieuchopstm reviewed (changes requested):
Looks nice! A few points that I think need to be addressed/checked:

pillo79 reviewed (commented):
Thanks everyone for your comments, will try to address all of them :slightly_smiling_face:

pillo79 committed:
llext: disable LL_EXTENSION_SYMBOL when not building an llext

The LL_EXTENSION_SYMBOL macro is used to export a symbol to the base
image. When CONFIG_LLEXT is not defined, or the file is being compiled
outside of an llext, the macro is not useful and would leave orphan
sections in the final image instead.

This patch adds the LL_EXTENSION_BUILD definition to the llext build
process, and uses it to stub out the symbol-defining macro when not
building an llext.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

mathieuchopstm commented on samples/subsys/llext/modules/README.rst:
```diff
@@ -0,0 +1,60 @@
+   :zephyr-app: samples/subsys/llext/modules
+   :board: qemu_xtensa
+   :goals: build run
+   :gen-args: -DCONFIG_LLEXT_HELLO_WORLD=y
```
Shouldn't this be `-T sample.llext.modules_disabled` instead?

mathieuchopstm commented on samples/subsys/llext/modules/README.rst:
```diff
@@ -0,0 +1,60 @@
+   :zephyr-app: samples/subsys/llext/modules
+   :board: qemu_xtensa
+   :goals: build run
+   :gen-args: -DCONFIG_LLEXT_HELLO_WORLD=m
```
Shouldn't this be `-T sample.llext.modules_enabled` instead?

mathieuchopstm reviewed (approved):
LGTM.

pillo79 commented:
Thanks @mathieuchopstm for making me think a bit more about what we are trying to explain here.

v2: 
* applied all review comments;
* added a way to detect that an llext build is in progress, and use it to stub out `LL_EXTENSION_SYMBOL` when the extension source file is included directly in the Zephyr build;
* renamed CONFIG_LLEXT_HELLO_WORLD to CONFIG_HELLO_WORLD_MODE and improved descriptions;
* made the test output different strings depending on the build type and cross-check them.

Ready for review/test as soon as CI turns this way... :slightly_smiling_face:

teburd commented:
@mathieuchopstm did you accept the invite? should have a green checkbox

mathieuchopstm commented:
> @mathieuchopstm did you accept the invite? should have a green checkbox

Pretty sure I did, I'm confused as well :thinking:
